File: crates/compiler/frontend/native/typescript/js/agent-program.ts
// Typed Agent Program authoring on the canonical five-op FrontendGraph surface.

import {
  GraphBuilder,
  canonicalAirJson,
  lower,
  verify,
  type Json,
} from "./graph-builder.ts";
import {
  type ProgramInstanceRef,
  type ProgramInvokeSpec,
  type ProgramNewSpec,
  type ProgramRef,
  programInvokeOperands,
  programNewOperands,
} from "./program-instance.ts";
import { StructuredTaskScope } from "./structured-task.ts";

export type HookBinding = {
  hook_id: string;
  scope: "agent" | "loop" | "node" | "model" | "capability";
  phase: "before" | "after";
  target_selector: string;
  declaration_order: number;
  handler_ref: string;
  handler_digest: string;
  input_type_ref: string;
  output_type_ref: string;
  return_mode: "observe" | "replace_result";
};

export type ContextEdge = {
  from_node: string;
  to_node: string;
  context_type_ref: string;
};

export type ImportedProgram = {
  program_ref: string;
  artifact_digest: string;
  entrypoint: string;
  target_agent_identity_requirement: string;
};

export class AgentProgram {
  readonly programId: string;
  readonly contextTypeRef: string | undefined;
  protected builder: GraphBuilder;

  constructor(opts: {
    program_id: string;
    entrypoint?: string;
    input_type_ref: string;
    output_type_ref: string;
    has_default_context?: boolean;
    context_type_ref?: string;
    source_language?: string;
  }) {
    this.programId = opts.program_id;
    this.contextTypeRef = opts.context_type_ref;
    this.builder = new GraphBuilder(opts.source_language ?? "typescript");
    this.builder.program({
      program_id: opts.program_id,
      entrypoint: opts.entrypoint ?? "run",
      input_type_ref: opts.input_type_ref,
      output_type_ref: opts.output_type_ref,
      has_default_context: opts.has_default_context ?? true,
      context_type_ref: opts.context_type_ref,
    });
  }

  importProgram(imported: ImportedProgram): this {
    this.builder.importProgram(imported);
    return this;
  }

  bindHook(hook: HookBinding): this {
    this.builder.hook(hook as Json);
    return this;
  }

  programNew(
    nodeId: string,
    spec?: ProgramNewSpec,
  ): { program: this; instance: ProgramInstanceRef } {
    this.builder.programNew(nodeId, spec === undefined ? undefined : programNewOperands(spec));
    const programRef = spec?.program_ref ?? nodeId;
    return { program: this, instance: { instance_node_id: nodeId, program_ref: programRef } };
  }

  programInvoke(
    nodeId: string,
    receiver: ProgramRef | ProgramInstanceRef,
    input?: Record<string, unknown>,
  ): this {
    const spec: ProgramInvokeSpec = { receiver, input };
    this.builder.programInvoke(nodeId, programInvokeOperands(spec));
    return this;
  }

  structuredTask(regionId: string): StructuredTaskScope {
    return new StructuredTaskScope(this, regionId);
  }

  contextFlow(edge: ContextEdge): this {
    this.builder.contextEdge(edge.from_node, edge.to_node, edge.context_type_ref);
    return this;
  }

  annotateRegion(regionId: string, annotation: string): this {
    this.builder.annotateRegion(regionId, annotation);
    return this;
  }

  buildGraph(): Json {
    return this.builder.build();
  }

  verifyGraph(): string | null {
    return verify(this.buildGraph());
  }

  lowerGraph(): Json {
    return lower(this.buildGraph());
  }

  canonicalAir(): string {
    return canonicalAirJson(this.buildGraph());
  }
}

File: crates/compiler/frontend/python/apxm_program/agent_program.py
"""Typed Agent Program authoring on the canonical five-op FrontendGraph surface."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Optional

from . import GraphBuilder, canonical_air_json, lower, verify
from .program_instance import ProgramInstanceRef, ProgramInvokeSpec, ProgramNewSpec, ProgramRef


@dataclass(frozen=True, slots=True)
class HookBinding:
    """Static Hook binding recorded on the FrontendGraph."""

    hook_id: str
    scope: str
    phase: str
    target_selector: str
    declaration_order: int
    handler_ref: str
    handler_digest: str
    input_type_ref: str
    output_type_ref: str
    return_mode: str

    def to_graph_fields(self) -> dict[str, Any]:
        return {
            "hook_id": self.hook_id,
            "scope": self.scope,
            "phase": self.phase,
            "target_selector": self.target_selector,
            "declaration_order": self.declaration_order,
            "handler_ref": self.handler_ref,
            "handler_digest": self.handler_digest,
            "input_type_ref": self.input_type_ref,
            "output_type_ref": self.output_type_ref,
            "return_mode": self.return_mode,
        }


@dataclass(frozen=True, slots=True)
class ContextEdge:
    """Explicit typed Context flow between two authored nodes."""

    from_node: str
    to_node: str
    context_type_ref: str


@dataclass(frozen=True, slots=True)
class ImportedProgram:
    """Digest-pinned import of an externally compiled specialist program."""

    program_ref: str
    artifact_digest: str
    entrypoint: str
    target_agent_identity_requirement: str


class AgentProgram:
    """Language-native wrapper over GraphBuilder for one authored Agent Program."""

    def __init__(
        self,
        *,
        program_id: str,
        entrypoint: str = "run",
        input_type_ref: str,
        output_type_ref: str,
        has_default_context: bool = True,
        context_type_ref: Optional[str] = None,
        source_language: str = "python",
    ) -> None:
        self.program_id = program_id
        self.context_type_ref = context_type_ref
        self._builder = GraphBuilder(source_language=source_language)
        self._builder.program(
            program_id=program_id,
            entrypoint=entrypoint,
            input_type_ref=input_type_ref,
            output_type_ref=output_type_ref,
            has_default_context=has_default_context,
            context_type_ref=context_type_ref,
        )

    @property
    def builder(self) -> GraphBuilder:
        return self._builder

    def import_program(self, imported: ImportedProgram) -> "AgentProgram":
        self._builder.import_program(
            program_ref=imported.program_ref,
            artifact_digest=imported.artifact_digest,
            entrypoint=imported.entrypoint,
            target_agent_identity_requirement=imported.target_agent_identity_requirement,
        )
        return self

    def bind_hook(self, hook: HookBinding) -> "AgentProgram":
        self._builder.hook(**hook.to_graph_fields())
        return self

    def program_new(
        self,
        node_id: str,
        *,
        spec: ProgramNewSpec | None = None,
        program_ref: str | None = None,
    ) -> tuple["AgentProgram", ProgramInstanceRef]:
        operands = spec.to_operands() if spec is not None else None
        if operands is None and program_ref is not None:
            operands = ProgramNewSpec(program_ref=program_ref).to_operands()
        self._builder.program_new(node_id, operands=operands)
        ref = program_ref or (spec.program_ref if spec is not None else node_id)
        return self, ProgramInstanceRef(instance_node_id=node_id, program_ref=ref)

    def program_invoke(
        self,
        node_id: str,
        *,
        receiver: ProgramRef | ProgramInstanceRef,
        input_payload: dict | None = None,
    ) -> "AgentProgram":
        spec = ProgramInvokeSpec(receiver=receiver, input_payload=input_payload)
        self._builder.program_invoke(node_id, operands=spec.to_operands())
        return self

    def structured_task(self, region_id: str) -> "StructuredTaskScope":
        from .structured_task import StructuredTaskScope

        return StructuredTaskScope(self, region_id)

    def context_flow(self, edge: ContextEdge) -> "AgentProgram":
        self._builder.context_edge(edge.from_node, edge.to_node, edge.context_type_ref)
        return self

    def annotate_region(self, region_id: str, annotation: str) -> "AgentProgram":
        self._builder.annotate_region(region_id, annotation)
        return self

    def build_graph(self) -> dict[str, Any]:
        return self._builder.build()

    def verify(self) -> Optional[str]:
        return verify(self.build_graph())

    def lower(self) -> dict[str, Any]:
        return lower(self.build_graph())

    def canonical_air_json(self) -> str:
        return canonical_air_json(self.build_graph())
